### ISSUE TITLE
Include collapsed subcollections of matching collections?

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#3838]
https://forums.zotero.org/discussion/112728/zotero-7-beta-navigate-to-sub-collections-from-filtered-collections

I guess we could consider this? A matching collection with non-matching subcollections would always be collapsed to start — or collapsed if already collapsed in the tree? — but you could expand manually.

I forget the details of how we're preserving the pre-filtered open state of the tree, and it's not totally clear how this would affect that — if you open a matching collection, does it stay open when you clear the search? — but maybe there's an easy way to do this.

**Comment by abaevbog** [comment]
This is certainly possible! While initially setting up the collection filter, I did take as an assumption that all collections in the tree are either matching the filter or have children that do, so some logic (especially when collections are renamed/added/removed) will likely need to be reworked.

>  A matching collection with non-matching subcollections would always be collapsed to start — or collapsed if already collapsed in the tree? — but you could expand manually.

Currently, when filtering starts, all collections from the tree are expanded. This is based on the assumption that all non-skipped collections are either matching or have children that do. If a matching collection has children that do not match, they are not included into the tree, so such collection has nothing to expand. So would it make sense to try to preserve this original behavior by expanding a collection row during filtering only if it has matching children? And then if you want to dig into a specific collection, you expand the relevant parents.

> if you open a matching collection, does it stay open when you clear the search?

When we clear the search, we do our best to restore the collectionTree to the state that it was in. If the selection was not changed, all collections are brought to the same `open` state as they were before. If a selection changed, in addition to that, we expand all parents of the selected collection, so that it is visible. I guess, we just keep this behavior? If the selected collection never changed, we'll just restore everything as it was. But if it did, it's probably fair to assume that the selected collection is what one was looking for, regardless if it was passing the filter or not, so we'll expand all it's parents and scroll to it?

The one question I am wondering about is how this affects keyboard nav? Currently, `arrowDown`/`arrowUp` skips over non-matching collections. We can keep this behavior without any issues, I think. But this would make it impossible to examine the non-matching subcollections of a matching collection.

**Comment by abaevbog** [comment]
https://forums.zotero.org/discussion/122492/show-nested-collections-when-filtering-collections

I wonder if including children or not could be a separate pref, so that those who do not want it (e.g. ones who have many collections, so their search results would be too cluttered) are not affected.

**Comment by dstillman** [comment]
I guess we could just start including them and wait for people to complain, and pref it if they do.

> Currently, arrowDown/arrowUp skips over non-matching collections. We can keep this behavior without any issues, I think. But this would make it impossible to examine the non-matching subcollections of a matching collection.

Well, we have this problem for non-matching ancestor collections now, right?

**Comment by dstillman** [comment]
(Sorry, I missed that we had a PR for this!)

**Comment by abaevbog** [comment]
> I guess we could just start including them and wait for people to complain, and pref it if they do.

Fair enough! I have that PR from a while back. I'll look at it again to rebase and make sure all still works. I'll try to refactor it so that, if we wanted to hide it behind a pref, it would not be too hard. 

> Well, we have this problem for non-matching ancestor collections now, right?

Correct. At the moment, grayed out non-matching ancestors are skipped during keyboard navigation. It made perfect sense for the general usability but I guess it doesn't really work for such specific situations.

**Comment by abaevbog** [comment]
I rebased and slightly reworked that older PR. Having the children of matches present made the initial approach of adding a hidden mock row to represent the currently selected row in itemTree too weird. Instead, it seems much simpler to clear the selection in collectionTree at all and restore it when the filter is cleared. If the selection is cleared, itemTree keeps displaying and using the initial `collectionTreeRow`.

Non-matching children are collapsed, so I don't actually expect this change to be an issue for anyone. Unless you expand the top-most matching collection, you never have to seen the children. The only significant difference is I think we have to restore default keyboard nav (instead of arrow up/down navigating only matches), otherwise there is no way at all to get to non-matching child matches via keyboard. I commented it out for now but maybe that's a candidate for the hidden pref if we ever add it.

The only thing I wasn't sure about is if the currently selected row is a non-matching child of a collection, should it get explicitly re-selected in the collectionTree if it's parent is expanded? I figured that no, so that you can expand/collapse collections without running into any selection changes but correct me if that feels off. For reference, this is what I mean:

https://github.com/user-attachments/assets/093eeaa4-c99c-49bc-afbe-dbe0cbfa9836
